### PR TITLE
[proxywrapper] Add new port

### PR DESCRIPTION
**Pull request by yuvalg-MSFT**

### CODE REVIEW

File: ports/proxywrapper/CONTROL [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Source: proxywrapper
+Version: 1.0.0
+Description: Provides a standalone wrapper around the libproxy library.
```
Comment (JackBoosY) [inline ports/proxywrapper/CONTROL]: Please add homepage with `Homepage: https://github.com/microsoft/proxy-wrapper`

File: ports/proxywrapper/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,20 @@
+include(vcpkg_common_functions)
```
Comment (JackBoosY) [inline ports/proxywrapper/portfile.cmake]: We no longer need to include it.

File: ports/proxywrapper/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+)
+
+vcpkg_install_cmake()
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(COPY ${SOURCE_PATH}/license.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/proxywrapper)
+file(RENAME ${CURRENT_PACKAGES_DIR}/share/proxywrapper/license.txt ${CURRENT_PACKAGES_DIR}/share/proxywrapper/copyright)
```
Comment (JackBoosY) [inline ports/proxywrapper/portfile.cmake]: Please replace these 2 lines with:
`file(INSTALL ${SOURCE_PATH}/license.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)`

File: ports/proxywrapper/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    HEAD_REF master
+)
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+)
+
+vcpkg_install_cmake()
+
```
Comment (JackBoosY) [inline ports/proxywrapper/portfile.cmake]: If you want to install pdb files, you need to add `vcpkg_copy_pdbs()` here.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/proxywrapper/portfile.cmake (hunk 4 above)
Comment (yuvalg-MSFT) [reply ports/proxywrapper/portfile.cmake]: Yes but this is intended to non-Windows only

**Comment by yuvalg-MSFT** [comment]
Currently this PR is pending as the dependency of libproxy doesn't exist on vcpkg. I'm trying to see if this can be mitigated but if this doesn't work I'll drop this PR

File: ports/proxywrapper/portfile.cmake (hunk 4 above)
Comment (JackBoosY) [reply ports/proxywrapper/portfile.cmake]: If this port support UNIX only, please add `vcpkg_fail_port_install(ON_TARGET "Windows")` in the top.

**Comment by JackBoosY** [comment]
Hi @yuvalg-MSFT, I'm adding libproxy in #8931.

**Comment by yuvalg-MSFT** [comment]
Hi, the port is available but its build fails, so it cannot be just added: build failure can be seen in config-x64-linux-dbg-err.log:

**Comment by JackBoosY** [comment]
Since CI system has problem, I will rerun the test.

**Comment by yuvalg-MSFT** [comment]
@jackboosy have you merged the libproxy? If so you can also merge the proxywrapper

**Comment by JackBoosY** [comment]
@yuvalg-MSFT Nope, we just fix the CI issue, I will restart this PR test after my PR merge.

**Comment by vicroms** [comment]
@yuvalg-MSFT 

I added a `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)` check to `proxywrapper`, as it does not export any symbols at all when building it as a shared library.

**Comment by yuvalg-MSFT** [comment]
@jackboosy cool. I assume everything is ready now. I haven't checked this yet but I'll uninstall all packages and download the wrapper uding vcpkg, expecting this to work and install libproxy automatically. I'll post here again if I run into issues

**Comment by yuvalg-MSFT** [comment]
@JackBoosY @vicroms just tried to check this and I see two major issues. First, after git pulling the vcpkg I don't see the proxywrapper project over there - only libproxy and libmodman projects. I thought it was merged too. Secondly when I tried to install libproxy, I see build errors. please refer to the attached screenshot. I tried doing vcpkg update, etc. but this still fails 
proxywrapper

**Comment by vicroms** [comment]
@yuvalg-MSFT  
It is not merged yet to the main repository, however if you pull the master branch of your fork it should be there.

The `libmodman` build failure is unexpected, it builds on CI. 
I got a successful install on my Linux machine (Ubuntu 16.04):

```
git clone https://github.com/yuvalg-MSFT/vcpkg.git yuvalg-vcpkg
cd yuvalg-vcpkg
./bootstrap.sh
./vcpkg install proxywrapper
```

Can you attach the output of `<vcpkg>/buildtrees/libmodman/install-x64-linux-dbg-out.log`?

**Comment by JackBoosY** [comment]
LGMT, thanks!

**Review by JackBoosY — APPROVED** [review]

**Comment by yuvalg-MSFT** [comment]
@vicroms Attached libmodman folder from <vcpkg>/build/trees
[libmodman.zip](https://github.com/microsoft/vcpkg/files/3898302/libmodman.zip)

**Comment by vicroms** [comment]
@yuvalg-MSFT 
I found this error in the logs: **`/usr/bin/ld: cannot find -lz`**.  Which means that `zlib` is not installed, are there any `libz.so` and `libz.so.1` libraries when you run `ldconfig -p | grep libz`?
 
In my case I have a system copy of `zlib` installed, and that's why the build did not fail for me. I'll add `zlib` as an explicit dependency on the port so that it gets installed by vcpkg.

**Comment by vicroms** [comment]
Apparently `zlib` is only required if you're building `libmodman`'s tests so I moved the tests to a non-default feature as the base libmodan library doesn't require `zlib`.

@yuvalg-MSFT have you had the chance to test the `proxywrapper` port following the instructions on   https://github.com/microsoft/vcpkg/pull/8916#issuecomment-558880790?

**Comment by yuvalg-MSFT** [comment]
@vicroms apparent we still see failures and zlib is installed

Here’s the log:

 

The following packages will be built and installed:

  * libmodman[core]:x64-linux

    libproxy[core]:x64-linux

Additional packages (*) will be modified to complete this operation.

Starting package 1/2: libmodman:x64-linux

Building package libmodman[core]:x64-linux...

-- Using cached /home/jimson/tools/vcpkg/downloads/libmodman-2.0.1.zip

-- Using source at /home/jimson/tools/vcpkg/buildtrees/libmodman/src/ibmodman-2-3a43f5031d

-- Configuring x64-linux-dbg

-- Configuring x64-linux-rel

-- Building x64-linux-dbg

CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:136 (message):

    Command failed: /home/jimson/tools/vcpkg/downloads/tools/cmake-3.14.0-linux/cmake-3.14.0-Linux-x86_64/bin/cmake --build . --config Debug --target install -- -v

    Working Directory: /home/jimson/tools/vcpkg/buildtrees/libmodman/x64-linux-dbg

    See logs for more information:

      /home/jimson/tools/vcpkg/buildtrees/libmodman/install-x64-linux-dbg-out.log

 

Call Stack (most recent call first):

  scripts/cmake/vcpkg_build_cmake.cmake:91 (vcpkg_execute_build_process)

  scripts/cmake/vcpkg_install_cmake.cmake:24 (vcpkg_build_cmake)

  ports/libmodman/portfile.cmake:29 (vcpkg_install_cmake)

  scripts/ports.cmake:94 (include)

 

 

Error: Building package libmodman:x64-linux failed with: BUILD_FAILED

Please ensure you're using the latest portfiles with `.\vcpkg update`, then

submit an issue at https://github.com/Microsoft/vcpkg/issues including:

  Package: libmodman:x64-linux

  Vcpkg version: 2019.09.12-unknownhash

**Comment by vicroms** [comment]
I suspect that the issue should be a different one now, with my latest commit `libmodman` should no longer try to link against `zlib`. Please attach the log file:
     
```
See logs for more information:
      /home/jimson/tools/vcpkg/buildtrees/libmodman/install-x64-linux-dbg-out.log
```

**Comment by vicroms** [comment]
I fixed the MacOS regressions for `libproxy`. Now the build is failing due to issues with the source code in `proxywrapper`:

```
[1/3] /Library/Developer/CommandLineTools/usr/bin/c++   -I/Users/vagrant/azure_agent/_work/2/s/installed/x64-osx/include -fPIC -g -MD -MT CMakeFiles/proxywrapper.dir/ProxyWrapper.cpp.o -MF CMakeFiles/proxywrapper.dir/ProxyWrapper.cpp.o.d -o CMakeFiles/proxywrapper.dir/ProxyWrapper.cpp.o -c /Users/vagrant/azure_agent/_work/2/s/buildtrees/proxywrapper/src/d1f03b59dc-72fbf198fb/ProxyWrapper.cpp
FAILED: CMakeFiles/proxywrapper.dir/ProxyWrapper.cpp.o 
/Library/Developer/CommandLineTools/usr/bin/c++   -I/Users/vagrant/azure_agent/_work/2/s/installed/x64-osx/include -fPIC -g -MD -MT CMakeFiles/proxywrapper.dir/ProxyWrapper.cpp.o -MF CMakeFiles/proxywrapper.dir/ProxyWrapper.cpp.o.d -o CMakeFiles/proxywrapper.dir/ProxyWrapper.cpp.o -c /Users/vagrant/azure_agent/_work/2/s/buildtrees/proxywrapper/src/d1f03b59dc-72fbf198fb/ProxyWrapper.cpp
/Users/vagrant/azure_agent/_work/2/s/buildtrees/proxywrapper/src/d1f03b59dc-72fbf198fb/ProxyWrapper.cpp:9:22: error: use of undeclared identifier 'nullptr'
    char** proxies = nullptr;
                     ^
/Users/vagrant/azure_agent/_work/2/s/buildtrees/proxywrapper/src/d1f03b59dc-72fbf198fb/ProxyWrapper.cpp:10:5: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
    auto proxyFactory = px_proxy_factory_new();
    ^
/Users/vagrant/azure_agent/_work/2/s/buildtrees/proxywrapper/src/d1f03b59dc-72fbf198fb/ProxyWrapper.cpp:11:25: error: use of undeclared identifier 'nullptr'
    if (proxyFactory != nullptr)
                        ^
/Users/vagrant/azure_agent/_work/2/s/buildtrees/proxywrapper/src/d1f03b59dc-72fbf198fb/ProxyWrapper.cpp:20:12: error: use of undeclared identifier 'nullptr'
    return nullptr;
           ^
1 warning and 3 errors generated.
ninja: build stopped: subcommand failed.
```

Looks like we need to add a flag to enable C++11 or latest features in ProxyWrapper's CMakeList.txt.

**Comment by vicroms** [comment]
All regressions have been addressed and the port now builds on all supported platforms (only fails on UWP).

It should be ready to merge.


@yuvalg-MSFT have you been able to find out why the build fails for you? I haven't been able to reproduce the issue on my end.

**Comment by JackBoosY** [comment]
@vicroms Thanks!

**Comment by yuvalg-MSFT** [comment]
@vicroms @JackBoosY 
sorry for the delay in my response. I still see two issues:

1. proxywrapper isn't part of the master branch. I did "git pull" and I still see the libproxy over there only.

2. building the libproxy still fails in my Ubuntu. I attached the error logs from libmodman

[error.txt](https://github.com/microsoft/vcpkg/files/3912569/error.txt)

**Comment by vicroms** [comment]
Hi @yuvalg-MSFT 

> 1. proxywrapper isn't part of the master branch. I did "git pull" and I still see the libproxy over there only.

We want to make sure that the port works before merging it to `vcpkg/master`. As you have reported some issues with building the libraries we haven't merged it yet. For now, all changes are only available on the `master` branch of your fork (https://github.com/yuvalg-msft/vcpkg), from where this PR was opened.

> 2. building the libproxy still fails in my Ubuntu because LZ is missing. after I did this "sudo apt-get install zlib1g-dev", then it worked. But I thought you were going to make it dependent so it wil linstall automatically before libproxy, no?

The changes to install `zlib`  before `libproxy` from **vcpkg** are only available on this PR, they are not yet merged into `master` either.

Hope this clarifies the current status.

Please let me know if you can build ProxyWrapper following these instructions:

```
git clone https://github.com/yuvalg-MSFT/vcpkg.git yuvalg-vcpkg
cd yuvalg-vcpkg
./bootstrap.sh
./vcpkg install proxywrapper
```

**Comment by yuvalg-MSFT** [comment]
@vicroms I tested and it works fine. so we can now merge everything

**Comment by vicroms** [comment]
Thanks for the PR!

**Pull request merged**